bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+
+        // the ringbuffer should be exactly as big as requested
+        if (buffer_frames_ != buffer_frames)
+          {
+            Bse::warning ("JACK driver: ring buffer size not correct: (buffer_frames_ != buffer_frames); (%u != %u)\n",
+                          buffer_frames_, buffer_frames);
+            error = Bse::Error::INTERNAL;
+          }
+        JDEBUG ("ringbuffer size = %.3fms", buffer_frames_ / double (mix_freq_) * 1000);
+
+        /* initialize output ringbuffer with silence
+         * this will prevent dropouts at initialization, when no data is there at all
+         */
+        vector<float>	  silence (output_ringbuffer_.get_total_n_frames());
+        vector<const float *> silence_buffers (output_ringbuffer_.get_n_channels());
+
+        fill (silence_buffers.begin(), silence_buffers.end(), &silence[0]);
+
+        uint frames_written = output_ringbuffer_.write (buffer_frames, &silence_buffers[0]);
+        if (frames_written != buffer_frames)
+          Bse::warning ("JACK driver: output silence init failed: (frames_written != jack->buffer_frames)\n");
+
+      }
+
+    /* activate */
+    if (error == 0)
+      {
+        jack_set_process_callback (jack_client_, static_process_callback, this);
+        jack_set_latency_callback (jack_client_, static_latency_callback, this);
+        jack_on_shutdown (jack_client_, static_shutdown_callback, this);
+
+        if (jack_activate (jack_client_) != 0)
+          error = Bse::Error::FILE_OPEN_FAILED;
+      }
+
+    /* connect ports */
+    if (error == 0) // we may want to make auto connect configurable (so it can be turned off)
+      {
+        std::map<std::string, DeviceDetails> devices = query_jack_devices (jack_client_);
+        std::map<std::string, DeviceDetails>::const_iterator di;
+
+        di = devices.find (devid_);
+        if (di != devices.end())
+          {
+            const DeviceDetails &details = di->second;
+
+            for (uint ch = 0; ch < n_channels_; ch++)
+              {
+                if (details.output_ports > ch)
+                  jack_connect (jack_client_, details.output_port_names[ch].c_str(),
+                                              jack_port_name (input_ports_[ch]));
+                if (details.input_ports > ch)
+                  jack_connect (jack_client_, jack_port_name (output_ports_[ch]),
+                                              details.input_port_names[ch].c_str());
+              }
+          }
+      }
+
+    /* setup PCM handle or shutdown */
+    if (error == 0)
+      {
+        flags_ |= Flags::OPENED;
+
+        uint dummy;
+        pcm_latency (&dummy, &dummy);   // debugging only: print latency values
+      }
+    else
+      {
+        disconnect_jack (jack_client_);
+        jack_client_ = nullptr;
+      }
+    JDEBUG ("JACK: opening PCM \"%s\" readupble=%d writable=%d: %s", devid_.c_str(), readable(), writable(), bse_error_blurb (error));
```

Comment:
Please make this "JACK: %s: ...", devid_, ...
And fix "readable" ;-)

Reply:
Fixed.